cpp/src/parquet/encryption/crypto_factory.cc

```diff
@@ -172,8 +199,8 @@ std::shared_ptr<FileDecryptionProperties> CryptoFactory::GetFileDecryptionProper
     const KmsConnectionConfig& kms_connection_config,
     const DecryptionConfiguration& decryption_config, const std::string& file_path,
     const std::shared_ptr<::arrow::fs::FileSystem>& file_system) {
-  auto key_retriever = std::make_shared<FileKeyUnwrapper>(
-      &key_toolkit_, kms_connection_config, decryption_config.cache_lifetime_seconds,
```

Comment:
I'm probably missing something, but why wouldn't change `FileKeyUnwrapper` to accept an owning pointer instead?

Reply:
That's what I initially looked at doing, but this caused issues with `KeyToolkit::RotateMasterKeys`, which needs to construct a `FileKeyUnwrapper` using a reference to itself, and I couldn't find a better workaround. There are some more details on this problem in the PR description.

Reply:
Maybe this could work if `KeyToolkit` could inherit from `std::enable_shared_from_this`, and we remove the public `KeyTookit` constructor and enforce that it can only be created inside a `shared_ptr`? I'd be concerned that this could be an annoying breaking change for some users though if anyone is using `KeyToolkit` directly rather than via a `CryptoFactory`.

Reply:
You can have `FileKeyUnwrapper` optionally take an owning pointer, but also leave the non-owning variant.

Reply:
I've changed this now, much nicer thanks